### ISSUE TITLE
Less Cryptic Table Syntax

### ISSUE THREAD

**Issue by sirinath** [asciidoctor/asciidoctor#1397]
Understanding and wrapping you head around something like the following is very difficult. It is possible to have a less cryptic table Kung Fu syntax.

```
[cols="e,m,^,>s", width="25%"]
|===
|1 >s|2 |3 |4
^|5 2.2+^.^|6 .3+<.>m|7
^|8
|9 2+>|10
|===
```

**Comment by wolandscat** [comment]
Personally I doubt it, unless the syntax becomes either much more wordy and/or structural (i.e. like XML or JSON). If the latter, we will end up just editing something like HTML th / td/ tr blocks.

Practically speaking, what I suspect most do is copy the last one, and hack changes into it from the Asciidoctor user manual.

The only real way around this is to have a GUI app / table builder like the HTML table builders on the web - that turn visual choices into syntax. In an integrated authoring environment based on an editor like Brackets, it's not hard to imagine this as a plug-in.

**Comment by lordofthejars** [comment]
but maybe we can create some alias like `e` is equal to `emphasis`  `m` to `monospace` so you can write something like:

```
[cols="emphasis,monospace,^,>s", width="25%"]
|===
|
|
|===
```

or even

|===
emphasis| blabla
|===

**Comment by wolandscat** [comment]
I'm not against it at all - I just consume whatever the current syntax definition is. But I can never remember it. Also - and this is the key thing in many parts of a raw syntax documentation system - how do you know what choices are there?

Consider what it's like using an editor like Brackets, that knows what a CSS says, and gives you auto-complete drop-downs while you edit HTML tags. Or even just programmers' auto-complete like in VS or Eclipse or whatever. The environment of the future for Asciidoc is actually a programmer-like environment, not a raw text authoring environment (once you are past inline text modifications, basic headers and lists, you need tooling).

Having said all that, I'd support your proposal ;-)

**Comment by lordofthejars** [comment]
:) well my proposal is the the sense of making document more readable in raw, so any users can see at least if something is in emphasis or monospace.

@mojavelinux any thoughts?

**Comment by wolandscat** [comment]
More _readable_ raw - no-one can disagree with that - to a point. But there is always a trade-off between efficient readability and structural comprehensibility - e.g. what you get with HTML tables (just sticking with tables for now). Ugly, loathsome XML style double tags. But if you know just th/td/tr and the colspan and rowspan tricks, you can easily read an HTML table.

The real question is: knowing how to _write_, which is different from knowing what you _read_. I find I can remember most of the basics now of Asciidoc(tor) markdown. But to get a doc to decent presentable form: that's a whole other ballgame - knowing all the attributes for every macro...and then there's the HTML stylesheet factory; PDF theme; custom syntax colourisers. Ok, in theory just the publishing guru is going to know those last ones.

Oh - and the #1 killer feature for authoring? For me it would be tool-based X-ref support. I have now have 6 logical docs, each of around 10 separate files, referencing 3 more references / biblio files in common. So there are those references, the references to figures, headings, other docs, external HTML etc. All the internal ones should be tool-supported one day - again I am seeing the Brackets editor (a new find for me) as the kind of tool that could support it.

Making plug-ins to do this kind of thing well relies on reliable syntax so that the plug-in can track Xrefs, find tables and so on.

Anyway, this is a bigger discussion. I'm finding that Asciidoctor's balance of brevity and understandability is pretty good. It really makes the job of maintainers easy, because they normally only have to make small content changes.

I may blog on all this soon :)

**Comment by lordofthejars** [comment]
Thanks :) yes sure AsiiDoc(tor) format is really awesome but there is always space for improving, and what I was suggesting is an alias not a substitution, but well let's see.
Alex.

**Comment by mojavelinux** [comment]
> It is possible to have a less cryptic table Kung Fu syntax.

It's extremely unlikely that you'd have a table that uses that much formatting at once. In fact, the example you cited is intentionally dense because its sole purpose is to show off all the possible formatting. I now question the usefulness of such an example because its more bewildering than helpful. It's really a test case more than an example.

I do agree that the colspecs are hard to remember. Even I have difficulty remembering and I wrote the parser for them. There are two problems I see:

* order of formatting marks matters
* the alignment marks are confusing (because the same symbols are used for horizontal and vertical)

I think we could fix a lot by removing the enforced ordering of formatting marks. Then, you can reason as follows:

* the specs are comma separated
* the numbers are relative widths
* you use letters to change the cell type
* you use arrows to control the alignment

That makes it easy to come up with the most common formatting:

```
cols=">1h,e2<"
```

I think the alignment arrows should be as follows:

* < for left
* > for right
* | for center (horizontally) (currently ^)
* ^ for top (currently <)
* _ for bottom (currently >)
* - for middle (vertically) (currently ^)

Notice that we use a different mark for every position. This is much better than using the same symbols for horizontal and vertical and having to remember the order and delimiter. So now we can do:

```
cols="1h>,e2|,a^"
```

I'm considering @lordofthejars' idea of allowing the names of the styles to be written longhand, but I don't think that's the biggest issue here. Even with full names, the order is still really hard to remember.

I'm not sure how to introduce this change and maintain backward compatibility. It may be a change we have to defer to UniDoc. Though, if we can parse the existing syntax and the new syntax side-by-side, then we can do it.

In a nutshell:

* remove order (thanks to unique marks)
* use the symbols for alignment proposed above
* allow style names to be written longhand (maybe)

**Comment by mojavelinux** [comment]
We could also use `~` for bottom to be consistent with subscript.

**Comment by wolandscat** [comment]
Here's a [realistic table](https://raw.githubusercontent.com/openEHR/specifications-AM/master/docs/AOM2/master00-amendment_record.adoc), for reference. [Rendered form](http://www.openehr.org/releases/AM/latest/docs/AOM2/AOM2.html#_amendment_record).

Another example, this time a [class text](https://raw.githubusercontent.com/openEHR/specifications-AM/master/docs/UML/classes/archetype.adoc), and its [Rendered form](http://www.openehr.org/releases/AM/latest/docs/AOM2/AOM2.html#_archetype_class).

I have no expectation of being able to remember the syntax needed here. So I live with copy, paste & modify, and hope for a visual tool one day;-)

**Comment by mojavelinux** [comment]
This syntax doesn't seem all that complex:

```
cols="1,6a,2,2"
```

The only confusing thing I can see is whether "a" goes before or after the "6", but that's something that my proposal addresses.

There are already visual tools emerging, so they will happen. But the idea of a visual tool should be to help commit the syntax to heart so that you don't need them. At least, that's my goal for AsciiDoc.

**Comment by wolandscat** [comment]
It's not so much the cols line (which I can more or less remember), it's putting together a whole table, including rows with merged cells, vertical cell merging, colouring of certain rows and cells, line breaks within cell content, quoting of the bar character, and so on. once you get past a purely regular simple table, it becomes hard to do from memory.

I'm not particularly complaining about this mind you - our tables were ultimately easy enough to implement, I just don't happen to think many people are going to be able to create anything more than the most trivial/regular table without having to look up some syntax. And I don't see any need or possibility of the syntax being simpler - it's as complex as it needs to be.

There's just no getting out of the much greater cognitive gap between the source representation of a table and its rendered form, compared to say a bullet list or heading in a Markdown style syntax (again, apart from the most trivial small tables, that can in fact be made to look visually like a table in the source file).

How about a table syntax cheat card inside Brackets and AsciidocFX?
